### ISSUE TITLE
SphereWaterInteraction and RegisterDynamicWavesInput do not work together with flow

### ISSUE THREAD

**Issue by klauskobald** [wave-harmonic/crest#1138]
### Current Behavior

SphereWaterInteraction and RegisterDynamicWavesInput do not create waves in the areas that have flow splines:
![image](https://github.com/wave-harmonic/crest/assets/278428/54b30e2c-c1ef-476c-9d19-133b50c73b98)
You can see it at the border where the flow starts.

### Expected Behavior

waves everywhere

### Unity Version

2021.3.31

### Crest Version

4.19

### Render Pipeline

Universal

### Editor or Standalone

Editor (Play Mode)

### Environment

```markdown
- OS: Mac 13.1
- GPU: Apple M1 Pro
- Graphics API: Metal
- Target Platform: Mac
```

**Comment by klauskobald** [comment]
Sorry! I found the problem.
I was scaling the object with the spline by a factor of 4, for testing. That screwsed up the thing.
Setting all to 1 solved the problem!

**Comment by klauskobald** [comment]
Unfortunately it is still not really solved. I made some tests:
This is a sphere moving at a speed of 6 thru not flowing water:
![image](https://github.com/wave-harmonic/crest/assets/278428/561a6ee3-3998-43d8-9172-3ea5d0faba36)

And this is a sphere sitting still in a water flow of 6
![image](https://github.com/wave-harmonic/crest/assets/278428/fbaf5973-7cfc-4e9a-a0c9-87d81e052ce9)

It generates way less waves in the moving water. I have put the same FFT wave settings onto the flow component.

Overall the waterinteractions produces way less waves inside a flow area.
Any advise?

**Another** thing I found out is that the SphereWaterIteraction.VelocityOffset works on the world speed of the object. This is wrong! It has to use the relative speed thru the water, right?

**Comment by klauskobald** [comment]
hm, I found in the code that spherewaterinteraction is actually using the relative waterspeed. So, I am even more confused now.

**Comment by klauskobald** [comment]
Better images:
Flow: OFF
![image](https://github.com/wave-harmonic/crest/assets/278428/5917cd86-f474-4ff5-bf10-dc2b10ceb9da)
Flow: ON
![image](https://github.com/wave-harmonic/crest/assets/278428/240b05d8-2af5-4b73-a7b8-2fba68192a44)

Same relative speed of 6.
Seems also to be a difference in the created foam.

**Comment by daleeidd** [comment]
Yeah, that doesn't seem right. I would think it would be stronger if anything but I could be wrong. If you disable the following line it does a bit to keep the interaction strength - can go even further by changing it to an add to increase the interaction strength:
https://github.com/wave-harmonic/crest/blob/92af05025c40f6ca22538eb6593a35df2c91cbc8/crest/Assets/Crest/Crest/Scripts/Interaction/SphereWaterInteraction.cs#L170

LMKWYT

> **Another** thing I found out is that the SphereWaterIteraction.VelocityOffset works on the world speed of the object. This is wrong! It has to use the relative speed thru the water, right?

That makes sense and will look into it.

**Comment by klauskobald** [comment]
Yes! That's it. Looks right to me! Please do not forget to change this for the next update!

**Comment by kyleli** [comment]
Unfortunately removing that line or adding the surfaceFlow vector to relativeVelocity creates a new issue where the interaction sphere effect begins to compound on itself over time when no input is applied. This is what happens when you remove the line with a flow of 9 and do not add any input to the spheres (just let it move with the flow).

https://github.com/wave-harmonic/crest/assets/1357336/d846664e-2513-46fe-bcf6-dc277e3c412a

**Comment by daleeidd** [comment]
Thanks for the additional input. I will look into it further and see if I can come up with something.

**Comment by daleeidd** [comment]
I had a harder look and the velocity calculations are correct.

I am starting to doubt this is a problem for SWI. A strong current can have a dampening effect on waves and I would suspect wakes too. If you have a propulsion system then normally it would have to work harder to fight the current and generate its own turbulence and maybe that is where the loss of SWI can be compensated by modelling those interactions. Furthermore, water spray is another area where this can be modelled.

There is an issue where wakes do not work very well at higher velocities and this can be caused by flow if going against it as it will increase the relative velocity, but a separate issue.

Happy to hear counterpoints.

**Comment by daleeidd** [comment]
Alright... so wakes are far more diminished when flow is not axis aligned. When axis aligned it is pretty close to expectations.

**Comment by kyleli** [comment]
> I had a harder look and the velocity calculations are correct.
> 
> There is an issue where wakes do not work very well at higher velocities which flow can contribute to and may be related in certain cases.
> 
> There is certainly a diminished wake in flow. I am not sure how SWI should behave in flow and whether this is correct or needs fixing. It may take a while before a conclusion on that.
> 
> For the time being I think the loss of wakes can be compensated by modelling the boat's propulsion system and further compensated with water spray as the boat will have to work harder against flow to maintain velocity.

I do actually agree with this, I was doing thinking as well and flow should definitely dampen the effect of SphereWaterInteractions, however I think the current iteration of flow is neutering this far too much. If we refer to even shallow rapids we can see a noticeable impact of the watercraft e.g. https://youtu.be/FW8c75CUBGU. Perhaps the impact of water interactions in flow should be a tunable setting on the flow side?

**Comment by daleeidd** [comment]
> e.g. in flows of 5 or above

Fastest river speed would be around 3MPS (in rare cases almost 5MPS). Flash flooding is around 7MPS.

This will take a while to solve properly as there is potentially various factors.

As for increasing the weight of SWI in flow that is a possible workaround. I made a branch you can try: [fix-weak-flow-swi-interaction](https://github.com/wave-harmonic/crest/tree/fix-weak-flow-swi-interaction)
